### PR TITLE
NTSL2++ sucessor to NTSL2+

### PR DESCRIPTION
**Pull request by Karolis2011**
This drastcly is meant to rework DM and Daemon code to allow further expansion and replacement of custom engine with JavaScript language based engine. 

Daemon PR - https://github.com/Aurorastation/ByondInterpretedLanguage/pull/7

Forums topic for discussion - https://forums.aurorastation.org/topic/14570-ntsl2-and-its-future/

Superseeds #8817 

Manditory check list:

- [x] Added subsystem for handling HTTP request handling
- [x] Added program types and there handling
- [x] Rewrote computer program to work with new system
- [x] New VueUI ui with code editor
- [x] Custom script files allow script data saving in computer filesystem
- [x] Rewrite NTSL2++ for dummies book.

Axed due to being too complicated and elaborate to implement with current systems:
- Add new UI to tcoms from scripting
- Test tcoms.

### CODE REVIEW

**Comment by Karolis2011** [comment]
!review

So now when I am open to reviews, here is little demo - https://cdn.discordapp.com/attachments/302201161044328450/739129615825764443/j9QeF2jg3p.mp4

File: code/_helpers/maths.dm [hunk 1]
```diff
@@ -139,3 +139,35 @@
...
+/proc/n_round(var/num)
+	if(isnum(num))
+		if(num-round(num)<0.5)
+			return round(num)
+		return n_ceil(num)
+
+// Clamps N between min and max
+/proc/n_clamp(var/num, var/min=-1, var/max=1)
+	if(isnum(num)&&isnum(min)&&isnum(max))
+		if(num<=min)
```
Comment (mikomyazaki) [inline code/_helpers/maths.dm:164]: Why not use clamp()?

File: code/modules/modular_computers/file_system/programs/generic/ntsl2_interpreter.dm [hunk 2]
```diff
@@ -1,100 +1,145 @@
...
-				files.Add(list(list(
-					"name" = F.filename,
-					"type" = F.filetype,
-					"size" = F.size,
-					"undeletable" = F.undeletable
-				)))
-		data["files"] = files
-
-	ui = SSnanoui.try_update_ui(user, src, ui_key, ui, data, force_open)
+/datum/computer_file/program/ntsl2_interpreter/ui_interact(mob/user as mob)
```
Comment (mikomyazaki) [inline code/modules/modular_computers/file_system/programs/generic/ntsl2_interpreter.dm:98]:
```suggestion
/datum/computer_file/program/ntsl2_interpreter/ui_interact(mob/user)
```

File: code/modules/modular_computers/file_system/programs/generic/ntsl2_interpreter.dm [hunk 3]
```diff
@@ -1,100 +1,145 @@
...
+	// Gather data for computer header
+	var/headerdata = get_header_data(data["_PC"])
+	if(headerdata)
+		data["_PC"] = headerdata
+		. = data
+
+	var/obj/item/computer_hardware/hard_drive/hdd = computer?.hard_drive
+	
+	if(is_running && istype(running))
+		VUEUI_SET_CHECK(data["mode"], "program", ., data)
```
Comment (mikomyazaki) [inline code/modules/modular_computers/file_system/programs/generic/ntsl2_interpreter.dm:123]:
```suggestion
		data["mode"] = "program"
```
We're returning data no matter what here anyway, so no need to VUEUI_SET_CHECK.

File: vueui/src/components/view/mcomputer/ntsl/list.vue [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+          <vui-button :params="{execute_file: f.name}">Execute</vui-button>
+          <vui-button :params="{edit_file: f.name}">Edit</vui-button>
+        </td>
+      </tr>
+    </table>
+  </div>
+</template>
+
+<script>
+import Utils from "../../../../utils"
```
Comment (mikomyazaki) [inline vueui/src/components/view/mcomputer/ntsl/list.vue:30]: Does this have to be a relative path?

**Comment by mikomyazaki** [comment]
Code & changes seem fine. Just need that guidebook & I'll write a few programs to test out various features when that exists.

In testing I couldn't make Term.write() actually output anywhere visible?

**Comment by Karolis2011** [comment]
!review

So yeah to make it work you need to run deamon, that is why @mikomyazaki didn't get output. Well, so far if deamon is not running execute button just does nothing.

Deamon is here https://github.com/Karolis2011/ByondInterpretedLanguage/tree/ntsl2-js if you trust me you can get compiled verson via Discord DM

**Review by skull132 — APPROVED** [review]

**Pull request merged**